model/model_training/custom_datasets/instruction.py

```diff
@@ -21,25 +22,65 @@
     "oa_wiki_qa_bart_10000row": "michaelthwan/oa_wiki_qa_bart_10000row",
     "oa_leet10k": "ehartford/oa_leet10k",
     "poem_instructions": "checkai/instruction-poems",
+    "oa_stackexchange": "donfu/oa-stackexchange",
+    "tell_a_joke": "mikegarts/oa_tell_a_joke_20000",
+    "wizardlm_70k": "ehartford/WizardLM_alpaca_evol_instruct_70k_unfiltered",
 }
 
 
 class InstructionDataset(Dataset):
-    def __init__(self, dataset, cache_dir, split, max_words=512, mode="sft"):
+    def __init__(self, dataset, cache_dir, split, mode="sft"):
         assert mode in ("sft", "rl")
         self.name = dataset
-        self.dataset = load_dataset(INSTRUCTION_DATASETS[dataset], cache_dir=cache_dir, split=split)
-        self.instruction_column = "INSTRUCTION" if dataset != "minimath" else "question"
-        self.response_column = "RESPONSE" if dataset != "minimath" else "answer"
-        self.max_words = max_words
+        self.mode = mode
+        if dataset == "minimath":
+            self.instruction_column = "question"
+            self.response_column = "answer"
+        elif dataset == "wizardlm_70k":
+            self.instruction_column = "instruction"
+            self.response_column = "output"
+        else:
+            self.instruction_column = "INSTRUCTION"
+            self.response_column = "RESPONSE"
+
+        ds = load_dataset(INSTRUCTION_DATASETS[dataset], cache_dir=cache_dir, split=split)
+        self.dataset = []
+        num_invalid = 0
+        for i in range(len(ds)):
+            data = ds[i]
+            if (
+                data[self.instruction_column] is not None
+                and len(data[self.instruction_column].strip()) > 0
+                and data[self.response_column] is not None
+                and len(data[self.response_column].strip()) > 0
+                and _filter_by_words(data[self.instruction_column])
+                and _filter_by_words(data[self.response_column])
+            ):
+                self.dataset.append(data)
+            else:
+                num_invalid += 1
+        if num_invalid > 0:
+            print(f"[Warning] {num_invalid} entries of {dataset} were invalid.")
 
     def __len__(self):
         return len(self.dataset)
 
     def __getitem__(self, idx) -> DatasetEntry:
         data = self.dataset[idx]
         lang = None
-        # these datasets have been found to have above 95% english sentences.
-        if self.name in ["grade_school_math_instructions"]:
+        # use "en" for datasets which have more than 95% English messages
```

Comment:
this will teach the model to sometimes (with a < 5% chance) generate a language other than english even though the language is specified as english. probably not an issue, but it would be best to filter the affected datasets or augment them with language attributes.

Reply:
I have roughly looked through all the datasets that have been listed here and I had not found any non-English entries. Maybe there are some, but I don't think that those other languages play an important role. I think Tobias used an automatic language detection tool to measure the percentage of English items in the datasets (and he initially added only `grade_shool_math_instructions`). I added the other ones. If you think that one of them is not (very) predominantly English let me know.

Reply:
i see, then it’s probably a high-effort low-payoff issue. just something to be aware of and to watch out for. if we notice that the model sometimes randomly generates different languages, this could be one possible cause

Reply:
Yes, I just checked with [langdetect](https://github.com/Mimino666/langdetect) and just found that one. +1 for @dvruette comment, we should totally be aware of that.